### ISSUE TITLE
UI Press - Build migration for UI Press Book content type

### ISSUE THREAD

**Issue by pyrello** [uiowa/uiowa#4662]
1108 items to migrate on the original site.

# Proposed solution
- [x] Write migration from UI Press Book content to Book content.
- [x] Account for change in how book types are handled. On the old site, there are discrete fields for all book type fields. On the new site, there is a paragraph bundle with generic fields.
- [x] Additionally account for the fact that ebook ownership license is also handled differently. Previously there was a price field for both types of ownership. Now, the ownership type is property of the book type. If there was price information for both types of ownership, then two book type bundles will need to be created to account for both of them.
- [x] We would like to do some DOM manipulation on the Reviews field: Find every instance of `<p></p>` tags and replace them with `<blockquote></blockquote>` tags. It seems like it might also be possible to parse each review (each is inside its own `<p></p>` tag in examples I examined) and figure out the part that is the attribution and format it correctly.
- [x] Add redirect migration.

**Comment by pyrello** [comment]
We would like to figure out a way of identifying source images that are too small so that we can generate a list for fixing them later.

**Comment by pyrello** [comment]
@dnkemper, @e-marie-w, and @makurj, I'm wondering if you can give me any insight into what the intention was with the following fields. I can't seem to figure out how or to what they should map.
* `field_book_art`
* `field_book_file_uploads`
* `field_book_pages`
* `field_book_trim_size`

**Comment by joewhitsitt** [comment]
going to see if i can push this forward

**Comment by e-marie-w** [comment]
- `field_book_art` = part of the `PP., plates, indices` field on d7 site
- `field_book_file_uploads` - new field they requested, nothing to map from old site
- `field_book_pages` - also part of the `PP., plates, indices` field on d7 site
- `field_book_trim_size` - also part of `PP., plates, indices` field on d7 site

...I did not think ahead to how difficult it might make migration when they asked for the `PP., plates, indices` field to be broken up into three separate fields. Not every book will have info for all three fields. As far as I can tell, the page number always has 'pages' after it, and the trim size 'inches' after the numbers. The art can say anything from figures, photographs, maps, charts, etc. Usually there's commas between the pages, art, and trim size, but not always. And sometimes the art field will have multiple commas (2 charts, 5 photographs, etc) so it's probably not good to use the comma as the delimiter for that reason too.

**Comment by e-marie-w** [comment]
Annnnd there's 1124 books, okay great. Uhhh we could try something where commas are the delimiter, and then if it has 'pages' the content goes to pages, 'inches' to trim size, and everything else to art? And something similar to how we output a list of everything with a small image size. Except this time make a list of every book that doesn't have any commas in that field so we know which will need to be fixed?

**Comment by joewhitsitt** [comment]
![huge mistake](https://media.giphy.com/media/A0FGCbbooHe1y/giphy.gif)

Update: avoiding PP., plates, indices right now and trying to work on the book type stuff.

Update update: I think I got the initial book type stuff working. Since type and isbn are required I got those wired up. There is a lot of if/else gotchas with some of the other fields like ebook perpetual vs 120 day.

**Comment by joewhitsitt** [comment]
Going to keep trying to move book type forward

**Comment by joewhitsitt** [comment]
Unrelated but wanted to capture: The book type has a sale start/end date that is evaluated whether to show or not during a preprocess. This is going to need the same 1 hour cache tag that we developed for https://github.com/uiowa/uiowa/pull/4934 or the editor is going to have to "refresh" this content the day of the start and end.

**Comment by pyrello** [comment]
> Unrelated but wanted to capture: The book type has a sale start/end date that is evaluated whether to show or not during a preprocess. This is going to need the same 1 hour cache tag that we developed for https://github.com/uiowa/uiowa/pull/4934 or the editor is going to have to "refresh" this content the day of the start and end.

Nice catch!

**Comment by pyrello** [comment]
A few more questions for you @e-marie-w:
* Can you confirm that all of the ISBN 10 fields are no longer needed and we don't need to capture that information anywhere?
* On the old site, there is one ebook sale price. Do you know if that applies to both the 120 day license and the perpetual license? Or just one of those?

**Comment by e-marie-w** [comment]
> On the old site, there is one ebook sale price. Do you know if that applies to both the 120 day license and the perpetual license? Or just one of those?

That is a good question. I can email the customer. 

> Can you confirm that all of the ISBN 10 fields are no longer needed and we don't need to capture that information anywhere?

I can confirm that they don't need to be migrated, while I'm emailing I'll check that it doesn't need to be captured at all. 

Update: sent email and cc'd everyone currently assigned to this.

**Comment by joewhitsitt** [comment]
>We would like to do some DOM manipulation on the Reviews field: Find every instance of <p></p> tags and replace them with <blockquote></blockquote> tags. It seems like it might also be possible to parse each review (each is inside its own <p></p> tag in examples I examined) and figure out the part that is the attribution and format it correctly.

@pyrello can you be more specific here? I have converted them to blockquote elements. The wysiwyg implementation of blockquote is different than the blockquote block (alignment, etc.) and the attribution seems a bit inconsistent unless we look for `—` and get it along with everything after and wrap it in something `<cite>`:

`—Elizabeth Taylor, “Editor's Choice,” <em>Chicago Tribune</em>`

`—<em>Library Journal</em>`

**Comment by pyrello** [comment]
@joewhitsitt When I wrote that up, I only looked at a few examples, so it may be the case that what I was hoping for is not possible based on the actual data. If at a minimum we can parse out the various quotes into using `<blockquote>` tags, then that is good enough for me.

**Comment by joewhitsitt** [comment]
>Test series mapping logic. This is a multi-value field, although I'm not sure if there are any books that use it.

doing a full migration run over lunch and then plan to review series mapping

Update: migration ran completely. the series wasn't being mapped appropriately because the field was limited to 1 value. Pushed a small update there.

**Comment by joewhitsitt** [comment]
I _think_ this is just waiting on a response regarding `PP., plates, indices`

**Comment by e-marie-w** [comment]
Heard back from customer and got confirmation we don't need to capture the 10 digit ISBNs. Also, the ebook sales price only applies to perpetual ownership ebooks.

**Comment by joewhitsitt** [comment]
>we don't need to capture the 10 digit ISBNs.

sounds good. that didn't make it into the migration, so no action needed there.

>the ebook sales price only applies to perpetual ownership ebooks.

This should be done. https://github.com/uiowa/uiowa/pull/4968/commits/0220b664ad3b1a00c0543905161bd3e9f771ca2f

Waiting for more clarity on PP., plates, indices

**Comment by joewhitsitt** [comment]
Looks like we got a response from the customer to migrate content as-is and they will separate things out over time:

- Create a new temp field that displays in the same way as the three separate fields?
- Map migration to one of those fields could work if the render labeling could be more generic.

Thoughts @pyrello?

**Comment by pyrello** [comment]
@joewhitsitt I replied to the customer to clarify, but I think your first solution above probably makes the most sense.

**Comment by pyrello** [comment]
I think I am going to work on this again.

**Comment by pyrello** [comment]
Just discovered another issue that will need to be dealt with. The old site used a filter (e.g. `[bg|1983]`) that pulled all the images from another field and displayed them in a text area. How to deal with these images has not been accounted for yet. I have emailed the customer to see how she would like to proceed.

**Comment by e-marie-w** [comment]
Is this how the file attachments end up between the reviews and the summary, like on [https://www.uipress.uiowa.edu/books/9781587298035/enchanted-by-prairie](https://www.uipress.uiowa.edu/books/9781587298035/enchanted-by-prairie)? I think they can be migrated to the Interior Images field.

**Comment by pyrello** [comment]
@e-marie-w Sorry I didn't see this before. Allison just confirmed what you were recommending. This is unblocked again.
